Formula/php@7.1.rb

```diff
@@ -0,0 +1,384 @@
+      --with-kerberos
+      --with-layout=GNU
+      --with-ldap
+      --with-ldap-sasl
+      --with-libedit
+      --with-libzip
+      --with-mcrypt=#{Formula["mcrypt"].opt_prefix}
+      --with-mhash
+      --with-mysql-sock=/tmp/mysql.sock
+      --with-mysqli=mysqlnd
+      --with-ndbm
+      --with-openssl=#{Formula["openssl"].opt_prefix}
+      --with-pdo-dblib=#{Formula["freetds"].opt_prefix}
+      --with-pdo-mysql=mysqlnd
+      --with-pdo-odbc=unixODBC,#{Formula["unixodbc"].opt_prefix}
+      --with-pdo-pgsql=#{Formula["libpq"].opt_prefix}
+      --with-pgsql=#{Formula["libpq"].opt_prefix}
+      --with-pic
+      --with-png-dir=#{Formula["libpng"].opt_prefix}
+      --with-pspell=#{Formula["aspell"].opt_prefix}
+      --with-unixODBC=#{Formula["unixodbc"].opt_prefix}
+      --with-webp-dir=#{Formula["webp"].opt_prefix}
+      --with-xmlrpc
+      --with-xsl
+      --with-zlib
+    ]
+
+    if MacOS.version < :lion
+      args << "--with-curl=#{Formula["curl"].opt_prefix}"
+    else
+      args << "--with-curl"
+    end
+
+    system "./configure", *args
+    system "make"
+    system "make", "install"
+
+    # Allow pecl to install outside of Cellar
+    extension_dir = Utils.popen_read("#{bin}/php-config --extension-dir").chomp
+    orig_ext_dir = File.basename(extension_dir)
+    inreplace bin/"php-config", lib/"php", prefix/"pecl"
+    inreplace "php.ini-development", "; extension_dir = \"./\"",
+      "extension_dir = \"#{HOMEBREW_PREFIX}/lib/php/pecl/#{orig_ext_dir}\""
+
+    config_files = {
+      "php.ini-development" => "php.ini",
+      "sapi/fpm/php-fpm.conf" => "php-fpm.conf",
+      "sapi/fpm/www.conf" => "php-fpm.d/www.conf",
+    }
+    config_files.each_value do |dst|
+      dst_default = config_path/"#{dst}.default"
+      rm dst_default if dst_default.exist?
+    end
+    config_path.install config_files
+
+    unless (var/"log/php-fpm.log").exist?
+      (var/"log").mkpath
+      touch var/"log/php-fpm.log"
+    end
+  end
+
+  def caveats
+    <<~EOS
+      To enable PHP in Apache add the following to httpd.conf and restart Apache:
+          LoadModule php7_module #{opt_lib}/httpd/modules/libphp7.so
+
+          <FilesMatch \.php$>
+              SetHandler application/x-httpd-php
+          </FilesMatch>
+
+      Finally, check DirectoryIndex includes index.php
+          DirectoryIndex index.php index.html
+
+      The php.ini and php-fpm.ini file can be found in:
+          #{etc}/php/#{php_version}/
+    EOS
+  end
+
+  def post_install
+    pear_prefix = share/"pear"
+    pear_files = %W[
+      #{pear_prefix}/.depdblock
+      #{pear_prefix}/.filemap
+      #{pear_prefix}/.depdb
+      #{pear_prefix}/.lock
+    ]
+
+    %W[
+      #{pear_prefix}/.channels
+      #{pear_prefix}/.channels/.alias
+    ].each do |f|
+      chmod 0755, f
+      pear_files.concat(Dir["#{f}/*"])
+    end
+
+    chmod 0644, pear_files
+
+    # Custom location for extensions installed via pecl
```

Comment:
@kabel we seem to be running in to some issues with the pecl linking between versions. As part of the versioned policies we also seem to be in breach.
```
A keg_only :versioned_formula should not post_install anything in the HOMEBREW_PREFIX that conflicts with or duplicates the non-versioned counterpart (or other versioned formulae). For example, a node@6 formula should not install its npm into HOMEBREW_PREFIX like the node formula does.
```
https://github.com/Homebrew/brew/blob/master/docs/Versions.md

Do you have any ideas if and how we could work around this ?

Reply:
I'd need more information about what kind of issues people are running into (issue link, console messages, logs).

Technically speaking, the post_install only fixes permissions and creates placeholder directories in `HOMEBREW_PREFIX`. It  does _configure_ (using a properly versioned system pear config file) `pear` to share a repository with all other counterparts. This is mainly because `pear` does not to engine checking when resolving dependencies, so only the package version is considered. In other words, the pear repository should be mostly engine version independent. Correct me if I'm wrong, but isn't this exactly what the node versions _are_ doing with npm. `pecl` _is_ configured with a versioned repository because it is engine dependent, so there should be no conflicts or duplicates. I think we were pretty careful to ensure there was no conflicts, but maybe we missed something.

In my opinion this is not a breach of the versioned policies.

I have all the versions installed all with a few extra extensions and pecl installed extensions and have not had any problems switching via `brew unlink php && brew link --force php@7.1`.

Reply:
>Correct me if I'm wrong, but isn't this exactly what the node versions are doing with npm.

They used to. The policy was added specifically to overturn that. See https://github.com/Homebrew/homebrew-core/pull/24349

Reply:
So the [current node functionality](https://github.com/Homebrew/homebrew-core/blob/37661dac3fa4e044b63d63348563a44c631553b7/Formula/node.rb#L103) is against the policy?

Reply:
It only applies to keg_only versioned formulae.

Reply:
I should have linked to one of [those](https://github.com/Homebrew/homebrew-core/blob/37661dac3fa4e044b63d63348563a44c631553b7/Formula/node%408.rb#L52) then.

Reply:
yes that is probably a bug